include/IECore/LRUCache.inl

```diff
@@ -38,117 +38,605 @@
+						// we take a write lock directly, because we know
+						// we'll need to write to the new item. When insertion
+						// found a pre-existing item we optimistically take
+						// just a read lock, because it is faster when
+						// many threads just need to read from the same
+						// cached item.
+						m_writable = inserted || mode == FindWritable || mode == InsertWritable;
+
+						if( m_itemLock.try_acquire( it->mutex, /* write = */ m_writable ) )
+						{
+							if( !m_writable && mode == Insert && it->cacheEntry.status() == LRUCache::Uncached )
+							{
+								// We found an old item that doesn't have
+								// a value. This can either be because it
+								// was erased but hasn't been popped yet,
+								// or because the item was too big to fit
+								// in the cache. Upgrade to writer status
+								// so it can be updated in get().
+								m_itemLock.upgrade_to_writer();
+								m_writable = true;
+							}
+							// Success!
+							m_item = &*it;
+							return true;
+						}
+						else
+						{
+							// The Item lock is held by another thread.
+							// We must release the Bin lock and retry. This
+							// avoids deadlock when the GetterFunction holding
+							// the Item lock calls back into the cache and tries to
+							// access another item in the same Bin.
+							binLock.release();
+						}
+					}
+				}
+
+				friend class Parallel;
+
+				const Item *m_item;
+				typename Item::Mutex::scoped_lock m_itemLock;
+				bool m_writable;
+
+		};
+
+		bool acquire( const Key &key, Handle &handle, AcquireMode mode )
+		{
+			return handle.acquire( bin( key ), key, mode );
+		}
+
+		void push( Handle &handle )
+		{
+			// Simply mark the item as having been used
+			// recently. We will then give it a second chance
+			// in pop(), so it will not be evicted immediately.
+			// We don't need the handle to be writable to write
+			// here, because `recentlyUsed` is atomic.
+			handle.m_item->recentlyUsed = true;
+		}
+
+		bool pop( Key &key, CacheEntry &cacheEntry )
+		{
+			// Popping works by iterating the map until an item
+			// that has not been recently used is found. We store
+			// the current iteration position as m_popIterator and
+			// protect it with m_popMutex, taking the position that
+			// it is sufficient for only one thread to be limiting
+			// cost at any given time.
+			PopMutex::scoped_lock lock;
+			if( !lock.try_acquire( m_popMutex ) )
+			{
+				return false;
+			}
+
+			Bin *bin = &m_bins[m_popBinIndex];
+			typename Bin::Mutex::scoped_lock binLock( bin->mutex );
+
+			typename Item::Mutex::scoped_lock itemLock;
+			while( true )
+			{
+				// If we're at the end of this bin, advance to
+				// the next non-empty one.
+				while( m_popIterator == bin->map.end() )
+				{
+					binLock.release();
+					m_popBinIndex = ( m_popBinIndex + 1 ) % m_bins.size();
+					bin = &m_bins[m_popBinIndex];
+					binLock.acquire( bin->mutex );
```

Comment:
Actually, `m_popMutex` just protects `m_popBinIndex` and `m_popIterator`. Here we're releasing one bin and acquiring the next bin, rather than releasing and re-acquiring a single bin. Another thread could definitely acquire either bin at some point during that process. This could lead us to conclude that all bins are empty when in fact some have been refilled after we left them, but in this case we just return and leave it to some other thread to do the work next time around...